Add back X11 legacy WM_NAME encodings

ctrlcctrlv opened the pull request:
Closes #4924.

Based on patches of the past, such as this work by James Cloos in July 2010: https://github.com/exg/rxvt-unicode/commit/d7d98751b7385416ad1694b5f1fde6c312ba20d5, as well as code comments in the Perl module X11::Protocol::WM (https://metacpan.org/pod/X11::Protocol::WM) and even the code to Xlib itself, which taught me that we should never have been using `XStoreName`, all it does is call `XChangeProperty`, hardcoded to `XA_STRING`!

What can I say, when the task is old school, the sources are too 😂

cc @morgant @slouken 

Sorry for any problems caused to y'all's release cycle, I would've never caused any on purpose. I still believe I'm right per the spec (as Qt once believed in 2014), but seven years later is still too late. I wonder who will be trying again in 2028? Not me again that's for sure.

ctrlcctrlv commented:
By the way if any future people (or even present people) care about my spec reasoning, here it is: _NET_WM_NAME is defined as the main window title property string in X11. WM_NAME is a _fallback_. It shouldn't be required to define a fallback, you're supposed to check the main one first! (“Google and Mozilla are wrong, I tell you!” shouted local insane programmer furry, to no one in particular.)

morgant commented:
@ctrlcctrlv This is essentially the same solution that I had come up with in https://github.com/morgant/SDL/commit/da639b64bec75cfe08af5e34bbfef3de89d6ec33, but I hadn't addressed `X11_MessageBoxCreateWindow()`. 

Since there are still a fair amount of other checks for `X_HAVE_UTF8_STRING`, I think forcing an error for that here is best left for a future refactor. Just my two cents and I'll leave that decision up to @slouken & @icculus.

ctrlcctrlv commented:
I didn't make it a hard error, just a log message:

```
     SDL_LogError(SDL_LOG_CATEGORY_VIDEO, "X server doesn't support UTF8_STRING, a feature introduced in 2000! This is likely to become a hard error in a future libSDL2.");
```

I maintained begrudgingly the places it's used, although I added some warnings to some places I thought were technically unsound and probably are at a high risk of failure under exotic locales (`X11_GetWindowTitle`).

Basically I split the difference between what I wanted to do (rip out all locale-specific code, force users to have `UTF8_STRING`, use `UTF8_STRING` to set both `WM_NAME` and `_NET_WM_NAME`) and what you wanted me to do (be conservative).

ctrlcctrlv commented:
If you're talking about the hard error I added when `XSupportsLocale(void)` doesn't return True, it's really unavoidable. If Xorg doesn't support its own locale, that's a hard error. That should _never_ happen, and checking for it is encouraged in the documentation of [`XmbTextListToTextProperty`](https://tronche.com/gui/x/xlib/ICC/client-to-window-manager/XmbTextListToTextProperty.html). It's what allows me to know that [this line](https://github.com/libsdl-org/SDL/pull/5029/files#diff-4d617e5c2011238b8ec8564c68b8a4bf491d9500c1cfc2d2660ceaa39d56c4f7R1885) is sane, otherwise it could be trying to tell me _either_ it's out of memory _or_ it doesn't recognize its own locale.

sezero commented:
Fails build because of `bool` return type of `SDL_X11_SetWindowTitle`: changing to plain `int` fixes it.

ctrlcctrlv commented:
Oh, okay. I'll change that, though all the builds did pass and it builds fine for me. Weird compiler/weird options? Anyway, not a big deal, I do now note that `bool` isn't used elsewhere.

sezero commented:
> Weird compiler/weird options?

Rather old system (centos-6.10, gcc4.4)

> Oh, okay. I'll change that, 

Thanks

slouken commented on src/video/x11/SDL_x11video.c:
```diff
@@ -460,6 +460,14 @@ X11_VideoInit(_THIS)
 #endif /* SDL_VIDEO_DRIVER_X11_XFIXES */
 
+#ifndef X_HAVE_UTF8_STRING
+    SDL_LogError(SDL_LOG_CATEGORY_VIDEO, "X server doesn't support UTF8_STRING, a feature introduced in 2000! This is likely to become a hard error in a future libSDL2.");
```
We shouldn't add an error message based on a preprocessor define. Instead this should be a build #warning

slouken commented on src/video/x11/SDL_x11video.c:
```diff
@@ -460,6 +460,14 @@ X11_VideoInit(_THIS)
+#endif
+
+    if (X11_XSupportsLocale() != True) {
+        return SDL_SetError("Current locale not supported by X server, cannot continue.");
```
We shouldn't introduce a video initialization failure here, even if it seems like it should always succeed. Instead we should check this when we're setting the title, and fail to do so there if necessary.

ctrlcctrlv replied:
Oh, all right. That makes sense, I appreciate your efforts in writing fault tolerant software. I write a lot of Rust and a lot of those folks value correctness above all else. (Meaning, crash/refuse to continue if anything is wrong.) Glad to know I don't have to code that way here, should we work together more beyond this patch. Fixed in 513e0eb.

slouken commented on src/video/x11/SDL_x11window.c:
```diff
@@ -716,9 +716,11 @@ X11_GetWindowTitle(_THIS, Window xwindow)
                     0L, 8192L, False, XA_STRING, &real_type, &real_format,
                     &items_read, &items_left, &propdata);
         if (status == Success && propdata) {
+            SDL_LogError(SDL_LOG_CATEGORY_VIDEO, "Failed to convert _WM_NAME title expecting UTF8! Title: %s", title);
```
In general we shouldn't add log messages to release code. You can make that LogDebug, or #ifdef some_debug_flag. If this is a fatal error, then we should return the message via SDL_SetError() and return an error code.

slouken commented on src/video/x11/SDL_x11window.c:
```diff
@@ -1888,6 +1871,39 @@ X11_FlashWindow(_THIS, SDL_Window * window, SDL_FlashOperation operation)
+    // we know this can't be a locale error as we checked X locale validity in X11_VideoInit
+    } else if (conv <= 0) {
+        SDL_LogError(SDL_LOG_CATEGORY_VIDEO, "X11 reporting it's out of memory");
+        return EXIT_FAILURE;
```
This function should return int and do this instead:
return SDL_OutOfMemory();

ctrlcctrlv replied:
Darn, should've known about that function. The fact I've made a single SDL patch before this one is showing, ain't it? :sweat_smile: Fixed in 513e0eb.

slouken commented on src/video/x11/SDL_x11window.c:
```diff
@@ -1888,6 +1871,39 @@ X11_FlashWindow(_THIS, SDL_Window * window, SDL_FlashOperation operation)
+        SDL_LogError(SDL_LOG_CATEGORY_VIDEO, "X11 reporting it's out of memory");
+        return EXIT_FAILURE;
+    } else { // conv >= 0
+        SDL_LogWarn(SDL_LOG_CATEGORY_VIDEO, "%d characters were not convertable to the current locale!", conv);
```
This function should return int and do this instead:
return SDL_SetError("%d characters were not convertable to the current locale!", conv);

ctrlcctrlv replied:
This is the only review comment I'm leery about and wish to discuss further. Because this is a fallback representation, I think we should be just a tad more fault tolerant regarding bad stuff in it. X11 will always put a string there, it's just, it'll report to you how many characters it _failed_ to encode. Note this:

> If the supplied text is not fully convertible to the specified encoding, the functions return the number of unconvertible characters. Each unconvertible character is converted to an implementation-defined and encoding-specific default string. 

X11 will put the `XDefaultString()` value anywhere that it couldn't do a requested decode. Consider, for example, a user on a ja_JP.SHIFT-JIS encoding Xorg (yes, in 2021, humor me for the sake of an example :laughing:). If they go on an Arabic webpage in an SDL2 web browser (yes my example just invented a competitor to Chrome, why not, it's causing this whole PR in the first place hehe), that'd be a failure to encode some characters, but shouldn't error out the whole title I don't think. I think it's good information to have if it starts going wrong on us, but there are legit reasons (in the early 2000's at least…) it might happen.

So I'd ask you to reconsider this request.

slouken commented on src/video/x11/SDL_x11window.c:
```diff
@@ -1888,6 +1871,39 @@ X11_FlashWindow(_THIS, SDL_Window * window, SDL_FlashOperation operation)
+#endif
+
+    X11_XFlush(display);
+    return EXIT_SUCCESS;
```
EXIT_SUCCESS and EXIT_FAILURE may not be defined. Use 0 for EXIT_SUCCESS.

sezero replied:
... and `-1` for EXIT_FAILURE

ctrlcctrlv replied:
Sorry about that, I thought these symbols were standard. Fixed in 513e0eb.

slouken commented on src/video/x11/SDL_x11window.c:
```diff
@@ -1871,22 +1872,25 @@ X11_FlashWindow(_THIS, SDL_Window * window, SDL_FlashOperation operation)
-    } else { // conv >= 0
-        SDL_LogWarn(SDL_LOG_CATEGORY_VIDEO, "%d characters were not convertable to the current locale!", conv);
-        return EXIT_FAILURE;
+    // we know this can't be a locale error as we checked X locale validity
```
Minor detail, we use /* */ style comments in SDL, to support really old compilers. :)

ctrlcctrlv replied:
On it, thanks, sorry for being a n00b, thank you for letting me clean up the mess I made, it's a tad embarrassing given that one of the things my career is based on is text encodings, I'm even a Unicode individual member and attend the meetings whenever I can make it :laughing: I just didn't expect any modern software to fail without `WM_NAME`, and the way this bug was found was such that I don't think any amount of testing would've lead me to find it. :(

ctrlcctrlv commented:
@slouken Just so you know I'd like to come back when the immediate crisis is over and do a complete review of SDL's text encoding story. Not just on X11, but primarily. Those issues I found with `X11_GetWindowTitle` I believe to be quite serious. There's no way that works on non-Unicode locales, it's guaranteed to garble. After the release let's talk about if we wanna require UTF8_STRING and what the story's gonna be. Since I use SDL2 in my MFEK font editor, this is very important to me and I'll dedicate the time necessary to get us into full modern text encoding compliance.

slouken commented:
Sure, that sounds good. Thanks for being patient with the needs of a new project. :)

ctrlcctrlv commented:
If you're in open source and lack patience boy do I have bad news for you :joy:

Remember how long the world waited for Wayland? I find it funny that people coming up now will have next to no idea that Ubuntu 14.10 was supposed to default to _Mir_ of all things because Canonical was scared we'd all be using Wayland by then and they wanted to have the chance to make their mark.

It's all voluntary association :) I wouldn't have it any other way.
